openslides_backend/migrations/mixins/deletion_mixin.py

```diff
@@ -79,130 +67,126 @@ def delete_update_by_schema(
-                    if not any(
-                        precursor
-                        for precursor in schema_part["precursors"]
-                        if deleted_instances[precursor] is None
-                    ):
-                        to_be_deleted_recursively: set = set()
-                        self.delete_collection(
-                            events,
-                            collection,
-                            schema_part,
-                            deleted_instances[collection],
-                            to_be_deleted_recursively,
-                            to_be_deleted,
-                            to_be_updated,
-                        )
-                        # safe all ids in deleted
-                        deleted_instances[collection] = to_be_deleted[collection]
-                        to_be_deleted[collection] = set()
-                        # delete same collection models recursively
-                        if to_be_deleted_recursively:
-                            self.delete_update_by_schema(
-                                {collection: to_be_deleted_recursively},
-                                deletion_schema,
-                                events,
-                            )
 
+        # set deletion root by finding statute related motions
+        # for collection, to_delete_ids in initial_deletions.items():
+        #     to_be_deleted[collection] = to_delete_ids
+        # delete until all collections in to_be_deleted have at least an empty list (means finished)
+        self._recursively_stage_for_deletion_and_rel_for_update(
+            initial_deletions,
+            deletion_schema,
+            to_be_deleted,
+            to_be_updated,
+        )
+        self.delete_all(to_be_deleted, events)
         # update lost references in bulk
         for collection, update_schema_part in update_schema.items():
             self.update_collection(
                 events,
                 collection,
                 update_schema_part,
                 to_be_updated[collection],
-                deleted_instances,
+                to_be_deleted,
             )
 
-    def is_finished(self, to_be_deleted: dict[str, set]) -> bool:
-        """Checks if all collections were handled for deletion."""
-        for collection in to_be_deleted.values():
-            if collection:
-                return False
-        return True
-
-    def delete_collection(
+    def _recursively_stage_for_deletion_and_rel_for_update(
         self,
-        events: list,
-        collection: str,
-        collection_delete_schema: dict[str, list],
-        collections_deleted_instance_ids: set[int] | None,
-        recursively_delete_ids: set[int],
+        initial_deletes: dict[str, set[int]],
+        delete_schema: MigrationDeletionSchema,
         to_be_deleted: dict[str, set[int]],
         to_be_updated: dict[str, dict[int, dict[str, list[int | str]]]],
```

Comment:
Minor point: In `shared/patterns.py` there is a type `Collection`, which is basically just a string. There may also a type `Id` defined somewhere in another file. Using this kind of verbose type, would make it easier for the reader to intuit, which part of your dict serves which function.

Reply:
You are right. This makes it much more readable. I used the types in `datastore.shared.typing`.